### PR TITLE
[MINOR] Async promise rejection message

### PR DESCRIPTION
**Pull request by michasherman**
| Q                | A
| ---------------- | ---
| Bug fix?         | no
| New feature?     | yes
| Breaking change? | no
| Deprecations?    | no
| Documentation?   | yes
| Tests added?     | yes

Added a new `test` api to support rejection message from async promise test callbacks. 

The new `test` api is defined as follows: `test ('someField', new Promise(...))`.

No breaking change is applied, so the standard `test` syntax will work as expected

### CODE REVIEW

File: src/core/test/index.js [hunk 1]
```diff
@@ -20,9 +20,13 @@ export const runAsync = (testPromise) => {
         }
     };
 
-    const fail = () => {
+    const fail = (rejectionMessage) => {
+        const message = typeof rejectionMessage === 'string'
```
Comment (ealush) [inline src/core/test/index.js]: IMO, statement should be the default - meaning that if the statement exists, we should always take it and not the rejection message. Not adding it is your way to communicate to passable that it should take the rejection message. Otherwise, why put it there in the first place?

Or maybe, what you think about is that your statement is a sort of a `catch all` message that will be used in case we can't get a rejection message?

File: src/core/test/spec.js [hunk 2]
```diff
@@ -104,64 +105,89 @@ describe('Test Passables "test" function', () => {
...
                         passable(lorem.word(), (test) => {
-                            test(f1, lorem.sentence(), new Promise((resolve) => done()));
+                            test(f1, lorem.sentence(), new Promise(() => done()));
                             test(lorem.word(), lorem.sentence(), noop);
                         });
                     });
 
+                    it('Should fail with rejection message when provided', (done) => {
+                        setTimeout(() => {
+                            expect(output.getErrors(f2)).to.deep.equal([rejectionMessage]);
```
Comment (ealush) [inline src/core/test/spec.js]: The test will be more resilient if you add a number of errors for this field, and check that the field includes this error.

File: docs/test/async.md [hunk 3]
```diff
@@ -16,17 +16,41 @@ test('name', 'must be unique', new Promise((resolve, reject) => {
...
 test('name',  'I fail', async () => Promise.reject());
+```
+
+## Rejecting with rejection message
+
+What if your promise can reject with different messages? No problem!
+You can reject the promise with your own message by passing it to the
+rejection callback.
+
+Notice that when using rejection messages we do not pass a `statement`
```
Comment (ealush) [inline docs/test/async.md]: That's not true according to your implementation. The correct wording according to your implementation is that "we do not *need* to pass statement", but we could, and it will be used as a fallback.

File: src/core/test/spec.js [hunk 4]
```diff
@@ -104,64 +105,89 @@ describe('Test Passables "test" function', () => {
...
                         f2 = lorem.word();
+                        f3 = lorem.word();
 
-                        output = passable(lorem.word(), (test, draft) => {
-                            test(f1, lorem.sentence(), new Promise((resolve, reject) => setImmediate(reject)));
-                            test(f2, lorem.sentence(), new Promise((resolve) => setTimeout(resolve)));
-                            test(f2, lorem.sentence(), new Promise((resolve) => setTimeout(resolve, 500)));
+                        output = passable(lorem.word(), (test) => {
+                            test(f1, lorem.sentence(), new Promise((_, reject) => setImmediate(reject)));
+                            test(f2, new Promise((_, reject) => setImmediate(() => reject(rejectionMessage))));
```
Comment (ealush) [inline src/core/test/spec.js]: At the moment you only added tests with either rejectionMessage or statement, not both. We don't test the case in which the rejectionMessage takes precedence (according to your current implementation).

File: src/core/test/index.js (hunk 1 above)
Comment (michasherman) [reply src/core/test/index.js]: > Or maybe, what you think about is that your statement is a sort of a catch all message that will be used in case we can't get a rejection message?

Exactly. I think it would be better if the rejectionMessage would always take precedence and the statement would serve as a `fallback` (for example in a case of an uncaught error).

Comment (ealush) [reply src/core/test/index.js]: I can relate to that logic. I can see some cases in which it would be unintuitive (but not counterintuitive), but these are edge cases and will happen due to human error.

File: docs/test/async.md [hunk 5]
```diff
@@ -16,17 +16,44 @@ test('name', 'must be unique', new Promise((resolve, reject) => {
...
+        });
 }));
 ```
 
 ## Returning a Promise / Async/Await
 
 ```js
 test('name',  'Should be unique', async () => {
-    const res = await doesUserExist(user)
+    const res = await doesUserExist(user);
```
Comment (ealush) [inline docs/test/async.md]: 🙈

File: docs/test/async.md [hunk 6]
```diff
@@ -16,17 +16,44 @@ test('name', 'must be unique', new Promise((resolve, reject) => {
...
 test('name',  'I fail', async () => Promise.reject());
+```
+
+## Rejecting with rejection message
+
+What if your promise can reject with different messages? No problem!
+You can reject the promise with your own message by passing it to the
+rejection callback.
+
+Notice that when using rejection messages we do not need pass `statement`
```
Comment (ealush) [inline docs/test/async.md]:
```suggestion
Notice that when using rejection messages we do not need to pass `statement`
```

File: src/core/test/index.js [hunk 7]
```diff
@@ -113,17 +117,41 @@ const register = (testFn) => {
...
+        [statement, testFn, severity] = args;
+    } else if (isTestFn(args[0])) {
+        [testFn, severity] = args;
+    }
+
     if (!testFn) {
         return;
     }
+
     if (typeof testFn.then === 'function' || typeof testFn === 'function') {
```
Comment (ealush) [inline src/core/test/index.js]: You can use `isTestFn` here.

File: src/core/test/spec.js [hunk 8]
```diff
@@ -87,81 +86,124 @@ describe('Test Passables "test" function', () => {
...
+                            test(f1, lorem.sentence(), new Promise(() => done()));
                             test(lorem.word(), lorem.sentence(), noop);
                         });
                     });
 
+                    it('Should fail with rejection message when provided', (done) => {
+                        setTimeout(() => {
+                            [f2, f3].forEach((field) => {
+                                expect(output.getErrors(field)).to.include(rejectionMessage);
+                                expect(output.getErrors(field)).not.to.include(field);
```
Comment (ealush) [inline src/core/test/spec.js]: What does the second line test exactly?

Comment (michasherman) [reply src/core/test/spec.js]: It's to check that the test statement is not part of the errors. I can remove it if you would prefer

File: src/core/test/index.js [hunk 9]
```diff
@@ -113,27 +117,49 @@ const register = (testFn) => {
...
+    }
+
+    return typeof testFn.then === 'function' || typeof testFn === 'function';
+};
+
 /**
  * The function used by the consumer
  * @param {String} fieldName name of the field to test against
- * @param {String} statement the message shown to the user in case of a failure
+ @param {String} [statement] the message shown to the user in case of a failure
```
Comment (ealush) [inline src/core/test/index.js]: Looks like you're missing an asterisk.
